reader/src/main/java/com/graphhopper/reader/osm/CustomOsmReader.java

```diff
@@ -0,0 +1,215 @@
+package com.graphhopper.reader.osm;
+
+import com.carrotsearch.hppc.LongIndexedContainer;
+import com.graphhopper.storage.*;
+import com.graphhopper.util.BitUtil;
+import com.graphhopper.util.EdgeIteratorState;
+import com.graphhopper.util.PointList;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public class CustomOsmReader extends OSMReader {
+    private static final BitUtil bitUtil = BitUtil.LITTLE;
+    private final DataAccess nodeMapping;
+    private final DataAccess edgeMapping;
+    private final DataAccess edgeAdjacentMapping;
+    private final DataAccess edgeBaseMapping;
+    private final NodeAccess nodeAccess;
+
+    public CustomOsmReader(GraphHopperStorage ghStorage) {
+        super(ghStorage);
+        Directory dir = ghStorage.getDirectory();
+        nodeMapping = dir.find("node_mapping");
+        nodeMapping.create(2000);
+        edgeMapping = dir.find("edge_mapping");
+        edgeMapping.create(1000);
+        edgeAdjacentMapping = dir.find("edge_adjacent_mapping");
+        edgeAdjacentMapping.create(1000);
+        edgeBaseMapping = dir.find("edge_base_mapping");
+        edgeBaseMapping.create(1000);
+        this.nodeAccess = ghStorage.getNodeAccess();
+    }
+
+    protected void storeOsmNodeID(int nodeId, long osmNodeId) {
+        long pointer = 8L * nodeId;
+        nodeMapping.ensureCapacity(pointer + 8L);
+        nodeMapping.setInt(pointer, bitUtil.getIntLow(osmNodeId));
+        nodeMapping.setInt(pointer + 4, bitUtil.getIntHigh(osmNodeId));
+    }
+
+    @Override
+    protected void storeOsmWayID(int edgeId, long osmWayId) {
+        super.storeOsmWayID(edgeId, osmWayId);
+        long pointer = 8L * edgeId;
+        edgeMapping.ensureCapacity(pointer + 8L);
+        edgeMapping.setInt(pointer, bitUtil.getIntLow(osmWayId));
+        edgeMapping.setInt(pointer + 4, bitUtil.getIntHigh(osmWayId));
+    }
+
+    protected void storeAdjacentNode(int edgeId, int adjacentNodeId) {
+        long pointer = 8L * edgeId;
+        edgeAdjacentMapping.ensureCapacity(pointer + 8L);
+        edgeAdjacentMapping.setInt(pointer, bitUtil.getIntLow(adjacentNodeId));
+        edgeAdjacentMapping.setInt(pointer + 4, bitUtil.getIntHigh(adjacentNodeId));
+    }
+
+    protected void storeBaseNode(int edgeId, int adjacentNodeId) {
+        long pointer = 8L * edgeId;
+        edgeBaseMapping.ensureCapacity(pointer + 8L);
+        edgeBaseMapping.setInt(pointer, bitUtil.getIntLow(adjacentNodeId));
+        edgeBaseMapping.setInt(pointer + 4, bitUtil.getIntHigh(adjacentNodeId));
+    }
+
+    /*
+    Copy of OSMReader.addOsmWay() that collects additional information about OSM node IDs
+    and the base/adjacent graphhopper nodes associated with each new graphhopper edge
+    */
+    @Override
+    Collection<EdgeIteratorState> addOSMWay(final LongIndexedContainer osmNodeIds, final IntsRef flags, final long wayOsmId) {
```

Comment:
This function was basically copied/pasted from the core graphhopper code - the only difference is the instances where we're calling the functions above this one, used for storing node IDs + base/adj node info for each edge